boards/bastwan: Support BastWAN Board

h-filzer opened the pull request:
### Contribution description

Add RIOT OS support for [BastWAN Board](https://github.com/ElectronicCats/BastWAN).
Board uses on rak4260 which itself uses on samr34 with integrated Lora.
Based on samr34 xplained pro board (`boards/samr34-xpro`) and adopted to RAK specific pin mappings. 

### Testing procedure

Fieldtested with SX127X module and semtech stack from `examples/lorawan`.
Tested all ADC channels.
UART, USB and GPIOS/LEDs are working as defined.


### Issues/PRs references

not known

benpicco reviewed (commented):
Looks good, CI has added some additional suggestions to the [file view](https://github.com/RIOT-OS/RIOT/pull/16181/files).

The page you linked mentions that this comes with the UF2 bootloader.
This is supported by RIOT, but it hasn't been tried on this MCU before.
Can you check if `samdx1-arduino-bootloader` works for this? (given you still have the bootloader on your board).

This will use bossa - if you need to use `uf2conv`, check out `boards/nrf52840-mdk-dongle`.

IMHO for simple things like this there is no need to preserve the entire file history, so you could drop b8de4268268b257771757ec7b4f91deb35046824. But add your own email address.
The author is responsible for maintaining the board after all ;)

h-filzer commented on boards/bastwan/board.c:
```diff
@@ -0,0 +1,53 @@
+    /* initialize the boards LED */
+    gpio_init(LED0_PIN, GPIO_OUT);
+
+#ifdef TX_SWITCH_PWR_PIN
```
Added comment in 272ec044bbe89e5528b25510bd0929b017f8c5f9

h-filzer commented on boards/bastwan/include/board.h:
```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2019 Mesotic SAS
```
see 272ec044bbe89e5528b25510bd0929b017f8c5f9

h-filzer commented on boards/bastwan/include/periph_conf.h:
```diff
@@ -0,0 +1,197 @@
+/**
+ * @brief   GCLK reference speed
+ */
+#define CLOCK_CORECLOCK     (48000000U)
```
see 272ec044bbe89e5528b25510bd0929b017f8c5f9

h-filzer commented on boards/bastwan/include/periph_conf.h:
```diff
@@ -0,0 +1,197 @@
+/* Timer 0 configuration */
+#define TIMER_0_CHANNELS    2
+#define TIMER_0_ISR         isr_tc0
+#define TIMER_NUMOF         (sizeof(timer_config) / sizeof(timer_config[0]))
```
fixed 272ec044bbe89e5528b25510bd0929b017f8c5f9

h-filzer commented on boards/bastwan/include/periph_conf.h:
```diff
@@ -0,0 +1,197 @@
+/* interrupt function name mapping */
+#define UART_0_ISR          isr_sercom3
+
+#define UART_NUMOF          (sizeof(uart_config) / sizeof(uart_config[0]))
```
fixed 272ec044bbe89e5528b25510bd0929b017f8c5f9

h-filzer commented on boards/bastwan/include/periph_conf.h:
```diff
@@ -0,0 +1,197 @@
+        .flags = I2C_FLAG_NONE
+    }
+};
+#define I2C_NUMOF          (sizeof(i2c_config) / sizeof(i2c_config[0]))
```
fixed 272ec044bbe89e5528b25510bd0929b017f8c5f9

h-filzer commented on boards/bastwan/Makefile.dep:
```diff
@@ -0,0 +1,7 @@
+
+ifneq (,$(filter saul_default,$(USEMODULE)))
+  USEMODULE += saul_gpio
+endif
```
will test uf2 later and submit new pr

benpicco replied:
Ok, but USB STDIO might be an easy addition - what do you think?

h-filzer replied:
I see the point that it might be easier for one to use cdc acm usb for stdio and debugging. I included your suggestion in my local build and did some power measurements. With cdc acm usb enabled the module power consumption increases about 40uA. Might be my fault somewhere but a qiuck look did not reveal a reason why the power consumption increases. 

Maybe you can give me a hint how to disable cdc acm in Makefile of project so that I can push the changes and have cdc acm disabled on demand? I am a starter in riot build and Makefiles :-)

benpicco replied:
Oh that's a good point!
It should be possible to only clock the USB peripheral if USB is plugged it, but this has not been implemented yet.
If you add an alternative stdio method in your application Makefile, `stdio_cdc_acm` will not be used.

e.g.

    USEMODULE += stdio_uart

h-filzer replied:
Works like a charm. Added `stdio_cdc_acm` in 76c0f028c0b51ec2932a1be0e4a6c2d22e5e0680

h-filzer commented:
Regarding the UF2 bootloader i have to disappoint you. No more UF2 bootloader installed. If testing the UF2 bootloader is valuable i will open another PR when I prepared one of the spare boards and tested.
All review advice's are implemented. Thanks for that!

kfessel commented:
> Nice one. I see that there's also the [BastBLE board](https://github.com/ElectronicCats/BastBLE) which is based on nrf52840, so should be easy to port to RIOT.

The RAK module series is a strange compilation every one of that modules has an other MCU

the RAK4600 even combines a nRF52832 (BLE) with an SX1276 (LORA)
(but it has a huge Ultra-Low Power Consumption 11.2μA in Sleep mode)

dylad commented on boards/bastwan/include/periph_conf.h:
```diff
@@ -0,0 +1,191 @@
+        .cts_pin = GPIO_UNDEF,
+#endif
+        .mux = GPIO_MUX_D,
+        .rx_pad = UART_PAD_RX_2,
+        .tx_pad = UART_PAD_TX_2,
```
Something's wrong here. You cannot map pad 2 to both RX and TX.

h-filzer replied:
Well spotted! Assigned correct pad : .rx_pad = UART_PAD_RX_3 in f1315e6b8024d6747af09943413e71b908d4098a

dylad commented on boards/bastwan/include/periph_conf.h:
```diff
@@ -0,0 +1,191 @@
+        .miso_pin = GPIO_PIN(PC, 19),
+        .mosi_pin = GPIO_PIN(PB, 30),
+        .clk_pin = GPIO_PIN(PC, 18),
+        .miso_mux = GPIO_MUX_F,
+        .mosi_mux = GPIO_MUX_F,
+        .clk_mux = GPIO_MUX_F,
```
SERCOM MUX on SAML21 are either C or D.

h-filzer replied:
spi got totally messed up. reassigned pins, pads and mux in f1315e6b8024d6747af09943413e71b908d4098a

h-filzer replied:
GPIO_MUX_F is the internal multiplexed signal for samr34<->semtech see [Datasheet Table 5-2](https://www.mouser.de/datasheet/2/268/SAMR34_R35_Low_Power_LoRa_Sub_GHz_SiP_Data_Sheet_D-1507280.pdf)
reminder to me: do not push after 11pm...

dylad replied:
I was looking at SAML21 datasheet, I completely forgot about this table. Sorry for the noise.

h-filzer replied:
Any hints and suggestions are valuable. I added the 2. spi_config to be able to connect spi periphs to the board so if you wanna x-check, plz..

dylad reviewed (dismissed):
Some peripherals misconfiguration.

benpicco commented:
Looks good to me, have your comments been addressed @dylad?

dylad commented:
> Looks good to me, have your comments been addressed @dylad?

Yes, this looks good.

@h-filzer could you run `dist/tools/compile_and_test_for_board/compile_and_test_for_board.py` with your board by any chance ?

Unless there is something weird, this PR should be in mergeable state.

h-filzer commented:
> > Looks good to me, have your comments been addressed @dylad?
> 
> Yes, this looks good.
> 
> @h-filzer could you run `dist/tools/compile_and_test_for_board/compile_and_test_for_board.py` with your board by any chance ?
> 
> Unless there is something weird, this PR should be in mergeable state.

@dylad I did run the tests. Took quite some time. Do you have an idea where i get all the test dependencies (python modules, pcap etc.) or how to rerun only failed tests? Anyway here is the result:

ERROR:bastwan:Tests failed: 16
Failures during compilation:
- [examples/gnrc_minimal](examples/gnrc_minimal/compilation.failed)
- [examples/lua_REPL](examples/lua_REPL/compilation.failed)
- [examples/suit_update](examples/suit_update/compilation.failed)
- [tests/periph_dma](tests/periph_dma/compilation.failed)
- [tests/periph_spi_dma](tests/periph_spi_dma/compilation.failed)
- [tests/pkg_cifra](tests/pkg_cifra/compilation.failed)
- [tests/pkg_emlearn](tests/pkg_emlearn/compilation.failed)
- [tests/pkg_nanopb](tests/pkg_nanopb/compilation.failed)
- [tests/suit_manifest](tests/suit_manifest/compilation.failed)
- [tests/unittests](tests/unittests/compilation.failed)

Failures during test:
- [examples/micropython](examples/micropython/test.failed)
- [tests/congure_test](tests/congure_test/test.failed)
- [tests/dbgpin](tests/dbgpin/test.failed)
- [tests/pbkdf2](tests/pbkdf2/test.failed)
- [tests/periph_timer_short_relative_set](tests/periph_timer_short_relative_set/test.failed)
- [tests/test_tools](tests/test_tools/test.failed)

dylad commented:
@h-filzer Thanks for this output ! @benpicco what do you think about this ?

You can re-run one test with `make BOARD=xxx -C tests/periph_dma flash test` for instance if you want to re-run `tests/periph_dma`. Be sure to not have your /dev/ttyACMx locked by another terminal before running your test.

h-filzer commented:
After adding missing packages here are the test and compile results for now:

Compiling

examples/gnrc_minimal - still searching for picolibc
examples/lua_REPL - ... `.bss' will not fit in region `ram'
examples/suit_update - error: 'public_key' undeclared (first use in this function)
- [x] tests/periph_dma 
- [x] tests/periph_spi_dma
tests/pkg_cifra - /pkg/cifra/src/tassert.h:27: error: "assert" redefined [-Werror]
tests/pkg_emlearn -  No module named 'emlearn'
- [x] tests/pkg_nanopb  
- [x] tests/suit_manifest
tests/unittests - `.text' will not fit in region `rom'

Tests

- [x] examples/micropython
tests/congure_test - AttributeError: 'NoneType' object has no attribute 'PM_TRAILING_COMMAS'
tests/dbgpin
- [x] tests/pbkdf2
tests/periph_timer_short_relative_set - interval 4 ok ERROR: too long delay, aborted after 1004 (TEST_MAX_DIFF=1000)
TEST FAILED
tests/test_tools -  assert retline.strip() == 'LOWERCASE'

The DMA Test passes now. In periph_conf.h for the SERCOM5 I had to remove 

```
#ifdef MODULE_PERIPH_DMA
        .tx_trigger = SERCOM5_DMAC_ID_TX,
        .rx_trigger = SERCOM5_DMAC_ID_RX,
#endif
```

in sercom5.h (line 123++)in samr34 has no index set?

```
#define SERCOM5_DMAC_ID_RX                   // Index of DMA RX trigger
#define SERCOM5_DMAC_ID_TX                   // Index of DMA TX trigger
#define SERCOM5_GCLK_ID_CORE        24      
#define SERCOM5_GCLK_ID_SLOW        23      
#define SERCOM5_INT_MSB             3       
#define SERCOM5_PMSB                3       
```

Any hints on fixin this ore leave it as is?

dylad commented:
> in sercom5.h (line 123++)in samr34 has no index set?

I guess this is because SERCOM5 dwells in PD0 while others SERCOM are in another power domain.
Moreover, Figure 4-2. System Block Diagram with MCU and Transceiver in SAMR34 datasheet seems to say that all SERCOM are DMA-capable except one so I think you can leave a small notice in the periph_conf.h to indicate why there is no DMA index for this SERCOM for future user.

dylad commented:
In any case I think you shouldn't remove this code
```
#ifdef MODULE_PERIPH_DMA
        .tx_trigger = SERCOM5_DMAC_ID_TX,
        .rx_trigger = SERCOM5_DMAC_ID_RX,
#endif
```
Because CI will complain.
So you can write a small notice in the periph_conf.h above the SPI configuration about this and leave it as it.
The proper fix for DMA should belong to another PR.

h-filzer commented:
> So you can write a small notice in the periph_conf.h above the SPI configuration about this and leave it as it.
> The proper fix for DMA should belong to another PR.
O.K. added code again and added small notice about nonfunctional DMA on SERCOM5 at the moment.

dylad commented:
> O.K. added code again and added small notice about nonfunctional DMA on SERCOM5 at the moment.

Great thanks ! Could you squash please ?
I think we're almost good. @benpicco how about you ?

benpicco commented:
CI is [not happy](https://ci.riot-os.org/RIOT-OS/RIOT/16181/e52c12c6d591a49918119c3c4a1cee9699b1574e/output/compile/tests/periph_dma/bastwan:gnu.txt) yet.

dylad commented:
Regarding CI and DMA for this board, I suggest to blacklist this board from the CI DMA test for now.
The issue is known and not really related to this PR. I can prepare a followup PR later to fix the situation so we can move forward with this one.

h-filzer commented:
> Regarding CI and DMA for this board, I suggest to blacklist this board from the CI DMA test for now.
> The issue is known and not really related to this PR. I can prepare a followup PR later to fix the situation so we can move forward with this one.

Shall I add  `TEST_ON_CI_BLACKLIST += bastwan` to `preriph_dma` and `periph_spi_dma` Makefiles including a small comment ?

benpicco commented on boards/bastwan/include/periph_conf.h:
```diff
@@ -0,0 +1,210 @@
+        .mosi_pad = SPI_PAD_MOSI_0_SCK_3,
+        .gclk_src = SAM0_GCLK_MAIN,
+#ifdef MODULE_PERIPH_DMA
+        .tx_trigger = SERCOM5_DMAC_ID_TX,
+        .rx_trigger = SERCOM5_DMAC_ID_RX,
```
So why not
```suggestion
        .tx_trigger = DMA_TRIGGER_DISABLED,
        .rx_trigger = DMA_TRIGGER_DISABLED,
```

dylad replied:
I was not aware of this option, definitely the best choice !

benpicco commented:
please squash!

dylad commented:
@h-filzer 
please add your board in BOARD_INSUFFICIENT_MEMORY variable within the failing tests/examples Makefile.ci file
- tests/bench_xtimer
- examples/lua_REPL

CI should be finally happy (you can squash again at the same time).

aabadie commented:
> please add your board in BOARD_INSUFFICIENT_MEMORY variable within the failing tests/examples Makefile.ci file

For `tests/bench_xtimer`, you should rather add  the board to the `LOW_MEMORY_BOARDS` variable:

https://github.com/RIOT-OS/RIOT/blob/9cad382dc80854b171afd4dbfef37721899f0547/tests/bench_xtimer/Makefile#L7

dylad reviewed (approved):
ACK.
Overall looks good to me. 
`dist/tools/compile_and_test_for_board/compile_and_test_for_board.py` didn't show any issues so I think this PR is good to go.
@h-filzer Thanks for your work !

dylad commented:
Here we go !
